interfaces/builtin/steam_support.go

```diff
@@ -0,0 +1,160 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+const steamSupportSummary = `allow Steam to configure pressure-vessel containers`
+
+const steamSupportBaseDeclarationPlugs = `
+  steam-support:
+    allow-installation: false
+    deny-auto-connection: true
+`
+
+const steamSupportBaseDeclarationSlots = `
+  steam-support:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const steamSupportConnectedPlugAppArmor = `
+# Allow pressure-vessel to set up its Bubblewrap sandbox.
+/sys/kernel/ r,
+@{PROC}/sys/kernel/overflowuid r,
+@{PROC}/sys/kernel/overflowgid r,
+@{PROC}/sys/kernel/sched_autogroup_enabled r,
+@{PROC}/pressure/io r,
+@{PROC}/@{pid}/uid_map rw,
+@{PROC}/@{pid}/gid_map rw,
+@{PROC}/@{pid}/setgroups rw,
+@{PROC}/@{pid}/mounts r,
+@{PROC}/@{pid}/mountinfo r,
+
+# Create and pivot to the intermediate root
+mount options=(rw, rslave) -> /,
+mount options=(rw, silent, rslave) -> /,
+mount fstype=tmpfs options=(rw, nosuid, nodev) tmpfs -> /tmp/,
+mount options=(rw, rbind) /tmp/newroot/ -> /tmp/newroot/,
+pivot_root oldroot=/tmp/oldroot/ /tmp/,
+
+# Set up sandbox in /newroot
+mount options=(rw, rbind) /oldroot/ -> /newroot/,
+mount options=(rw, rbind) /oldroot/dev/ -> /newroot/dev/,
+mount options=(rw, rbind) /oldroot/etc/ -> /newroot/etc/,
+mount options=(rw, rbind) /oldroot/proc/ -> /newroot/proc/,
+mount options=(rw, rbind) /oldroot/sys/ -> /newroot/sys/,
+mount options=(rw, rbind) /oldroot/tmp/ -> /newroot/tmp/,
+mount options=(rw, rbind) /oldroot/var/ -> /newroot/var/,
+mount options=(rw, rbind) /oldroot/var/tmp/ -> /newroot/var/tmp/,
+mount options=(rw, rbind) /oldroot/usr/ -> /newroot/run/host/usr/,
+mount options=(rw, rbind) /oldroot/etc/ -> /newroot/run/host/etc/,
+mount options=(rw, rbind) /oldroot/usr/lib/os-release -> /newroot/run/host/os-release,
+
+# Bubblewrap performs remounts on directories it binds under /newroot
+# to fix up the options (since options other than MS_REC are ignored
+# when performing a bind mount). Ideally we'd write a rule that
+# requires the remount option in combination with any other, but
+# AppArmor doesn't currently support that.
+mount options in (rw, ro, nosuid, nodev, noexec, remount, bind, silent, relatime) -> /newroot/{,**},
```

Comment:
Is it possible to write another eleven rules like the above, with smaller subsets of these?

Reply:
The mount calls this is intended to cover are from this code:

https://github.com/containers/bubblewrap/blob/main/bind-mount.c

It is working around the fact that calling mount() with MS_BIND causes it to ignore all flags passed other than MS_REC. So it is essentially doing the following:
1. call `mount(..., MS_BIND|MS_REC, ...)` to perform the mount.
2. parse /proc/self/mountinfo to discover the new mounts that have been created, and decode the mount flags used for each mount.
3. for each new mount, bitwise or some additional flags to the existing flags. If the result is different, perform a `mount(..., MS_SILENT|MS_BIND|MS_REMOUNT|new_flags,...)` call to update the flags.

So the exact combination of flags will depend on the existing mounts on the system, which could include however the user has configured their host sytem, as that gets exposed through `/var/lib/snapd/hostfs`. So I think it'd be somewhat more than 12 possibilities if avoiding the `in` syntax.

I think the full enumeration would be the cross product of:
1. `silent, bind, remount, nosuid`: always set
2. `nodev` or nothing
3. `ro` or `rw`
4. `noexec` or nothing
5. `noatime` or `relatime` or nothing
6. `nodiratime` or nothing

So I think there are 48 possibilities. Does that sound about right?

Reply:
Oh, yikes, that code is far more dynamic than I expected. I thought I was going to see an array of a dozen directories plus their flags in the sources... Feel free to ignore this entirely, then. Thanks for the explanation.